src/package-operations.ts

```diff
@@ -9,97 +9,223 @@ import {
+      intermediatePackageMetadata,
+      changedPackages,
+      versionSyncStrategy,
+    ),
+    changedPackages,
+  ];
 }
 
 /**
- * @param allPackages - The metadata of all packages in the monorepo.
- * @param synchronizeVersions - Whether to synchronize the versions of all
- * packages.
- * @param tags - All tags for the release's base git branch.
- * @returns The names of the packages to update.
+ * Given the intermediate equivalent, gets the _complete_ metadata for all
+ * packages in the monorepo, which amounts to a boolean for each package
+ * indicating whether it should be updated. This is computed by analyzing the
+ * monorepo's dependency graph.
+ *
+ * @param intermediatePackageMetadata - The intermediate package metadata of all
+ * monorepo packages.
+ * @param changedPackages - The set of packages that have changed.
+ * @param versionSyncStrategy - The version synchronization strategy.
+ * @returns The complete package metadata for all packages in the monorepo.
  */
-export async function getPackagesToUpdate(
-  allPackages: Record<string, PackageMetadata>,
-  synchronizeVersions: boolean,
-  tags: ReadonlySet<string>,
-): Promise<ReadonlySet<string>> {
-  // In order to synchronize versions, we must update every package.
-  if (synchronizeVersions) {
-    return new Set(Object.keys(allPackages));
-  }
+export function getCompletePackageMetadata(
+  intermediatePackageMetadata: ReadonlyMap<string, IntermediatePackageMetadata>,
+  changedPackages: ReadonlySet<string>,
+  versionSyncStrategy: VersionSynchronizationStrategies,
+): ReadonlyMap<string, PackageMetadata> {
+  const dependencyGraph = getMonorepoDependencyGraph(
+    intermediatePackageMetadata,
+  );
 
-  // If we're not synchronizing versions, we only update changed packages.
-  const shouldBeUpdated: Set<string> = new Set();
-  // We use a for-loop here instead of Promise.all because didPackageChange
-  // must be called serially.
-  for (const packageName of Object.keys(allPackages)) {
-    if (await didPackageChange(tags, allPackages[packageName])) {
-      shouldBeUpdated.add(packageName);
-    }
-  }
+  // This mutable set will be modified as the dependency graph is traversed.
+  const packagesToUpdate = new Set(changedPackages.values());
 
-  if (shouldBeUpdated.size === 0) {
-    throw new Error(`There are no packages to update.`);
-  }
-  return shouldBeUpdated;
+  /**
+   * Sidebar: Topological order
+   *
+   * "...a topological sort or topological ordering of a directed graph is a
+   * linear ordering of its vertices such that for every directed edge uv from
+   * vertex u to vertex v, u comes before v in the ordering."
+   * Ref: https://en.wikipedia.org/wiki/Topological_sorting
+   *
+   * In our usage, there is a directed edge from PackageA to PackageB if
+   * PackageA lists PackageB in any of the dependency fields in its
+   * `package.json`.
+   *
+   * `overallOrder` is a convention of the `dependency-graph` library, which by
```

Comment:
Is this true? 

The README for `dependency-graph` has the following example:

```
graph.addDependency('a', 'b');
graph.addDependency('b', 'c');

graph.dependenciesOf('a'); // ['c', 'b']
graph.dependenciesOf('b'); // ['c']
graph.dependantsOf('c'); // ['a', 'b']

graph.overallOrder(); // ['c', 'b', 'a']
```

So, `b` is a direct dependency of `a`, and `c` is a direct dependency of `b`.

When determining how to go through this list, if we view these as tasks as in the example provided by the [Wikipedia article on topological sorting](https://en.wikipedia.org/wiki/Topological_sorting#Examples), if `a` depends on `b`, then we should "complete" (or visit) `b` before `a`, and if `b` depends on `c`, we should "complete" (or visit) `c` before `b`. So the final order is `c`, `b`, `a`.

Reversing this order would mean that dependents would get visited before dependencies. Is that what we want?

Reply:
It's possible that one of two things is true:

a. in trying to visualize the graph in my head, the arrows are reversed
b. I am forgetting the difference between dependents and dependencies

But essentially the goal with all of this is that if a package is updated, resulting in a version bump of some sort, anything referring to that package (i.e. anything with that package in its dependencies, i.e. any of its dependents) should get bumped too. Right?